### PR TITLE
SPV-838(feat) AdminAPI

### PR DESCRIPTION
**Pull request by ac4ch**
# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: actions/admin/access_keys_old.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+// @Success		200 {number} int64 "Count of access keys"
+// @Failure		400	"Bad request - Error while parsing CountAccessKeys from request body"
+// @Failure 	500	"Internal Server Error - Error while fetching count of access keys"
+// @Router		/v1/admin/access-keys/count [post]
+// @Security	x-auth-xpub
+func accessKeysCount(c *gin.Context, _ *reqctx.AdminContext) {
+	logger := reqctx.Logger(c)
+	var reqParams filter.AdminCountAccessKeys
+	if err := c.Bind(&reqParams); err != nil {
+		spverrors.ErrorResponse(c, spverrors.ErrCannotBindRequest, logger)
```
Comment (dzolt-4chain) [inline actions/admin/access_keys_old.go:71]: In some places you return ErrorResponse with spvErrors unwrapped and in other you also use WithTrace(err), maybe we could also add with trace in these places?

File: actions/admin/contact.go [hunk 2]
```diff
@@ -20,54 +21,55 @@ import (
...
 func contactsSearch(c *gin.Context, _ *reqctx.AdminContext) {
 	logger := reqctx.Logger(c)
 	engine := reqctx.Engine(c)
-	var reqParams filter.SearchContacts
-	if err := c.Bind(&reqParams); err != nil {
-		spverrors.ErrorResponse(c, spverrors.ErrCannotBindRequest, logger)
+
+	searchParams, err := query.ParseSearchParams[filter.ContactFilter](c)
+	if err != nil {
+		spverrors.ErrorResponse(c, spverrors.ErrCannotParseQueryParams, logger)
```
Comment (dzolt-4chain) [inline actions/admin/contact.go:35]: same here about WithTrace

File: actions/admin/contact_old.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+// @Failure		400	"Bad request - Error while parsing SearchContacts from request body"
+// @Failure 	500	"Internal server error - Error while searching for contacts"
+// @Router		/v1/admin/contact/search [post]
+// @Security	x-auth-xpub
+func contactsSearchOld(c *gin.Context, _ *reqctx.AdminContext) {
+	logger := reqctx.Logger(c)
+	engine := reqctx.Engine(c)
+	var reqParams filter.SearchContacts
+	if err := c.Bind(&reqParams); err != nil {
+		spverrors.ErrorResponse(c, spverrors.ErrCannotBindRequest, logger)
```
Comment (dzolt-4chain) [inline actions/admin/contact_old.go:34]: ☝️

File: actions/admin/contact_old.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+		return
+	}
+
+	conditions, err := reqParams.Conditions.ToDbConditions()
+	if err != nil {
+		spverrors.ErrorResponse(c, spverrors.ErrCannotBindRequest, logger)
```
Comment (dzolt-4chain) [inline actions/admin/contact_old.go:40]: ☝️

File: actions/admin/destinations.go [hunk 5]
```diff
@@ -13,6 +13,7 @@ import (
 
 // destinationsSearch will fetch a list of destinations filtered by metadata
 // Search for destinations filtering by metadata godoc
+// @DDeprecatedRouter /v1/admin/destinations [post]
```
Comment (dzolt-4chain) [inline actions/admin/destinations.go:16]: Double D

File: actions/admin/models.go [hunk 6]
```diff
@@ -38,11 +38,18 @@ type CreateXpub struct {
 	Key string `json:"key" example:"xpub661MyMwAqRbcGpZVrSHU..."`
 }
 
+// CreateUser is the model for creating a user
+type CreateUser struct {
+	// Accepts a JSON object for embedding custom metadata, enabling arbitrary additional information to be associated with the resource
+	Metadata engine.Metadata `json:"metadata" swaggertype:"object,string" example:"key:value,key2:value2"`
+	// The xpub key
+	Key string `json:"key" example:"xpub661MyMwAqRbcGpZVrSHU..."`
```
Comment (dzolt-4chain) [inline actions/admin/models.go:46]: Maybe we could rename it to be something like XPubKey or something indicating that it is actually a public key and not a private one? WDYT?

File: engine/action_paymails.go [hunk 7]
```diff
@@ -23,6 +23,20 @@ func (c *Client) GetPaymailAddress(ctx context.Context, address string, opts ...
...
 }
 
+// GetPaymailAddressByID will get a paymail address model
+func (c *Client) GetPaymailAddressByID(ctx context.Context, id string, opts ...ModelOps) (*PaymailAddress, error) {
+
+	// Get the paymail address
+	paymailAddress, err := getPaymailAddressByID(ctx, id, append(opts, c.DefaultModelOptions()...)...)
+	if err != nil {
+		return nil, err
+	} else if paymailAddress == nil {
```
Comment (dzolt-4chain) [inline engine/action_paymails.go:33]: just a nit pick but we can delete the else here 😅

File: actions/admin/models.go (hunk 6 above)
Comment (ac4ch) [reply actions/admin/models.go:46]: ok you see i forgot to remove it. `CreateUser`

File: actions/admin/access_keys_old.go (hunk 1 above)
Comment (nonus25) [reply actions/admin/access_keys_old.go:71]: this is old code, so i left untouched, just renamed the file to old.

File: actions/admin/contact.go (hunk 2 above)
Comment (nonus25) [reply actions/admin/contact.go:35]: fixed

File: actions/admin/contact_old.go (hunk 3 above)
Comment (nonus25) [reply actions/admin/contact_old.go:34]: same old code, so been moved to _old.go and i left untouched.

File: actions/admin/contact_old.go (hunk 4 above)
Comment (nonus25) [reply actions/admin/contact_old.go:40]: same as above

File: actions/admin/destinations.go (hunk 5 above)
Comment (nonus25) [reply actions/admin/destinations.go:16]: corrected

File: engine/action_paymails.go (hunk 7 above)
Comment (nonus25) [reply engine/action_paymails.go:33]: ok i copied this method, but u right this logic isn't looking great :)

File: actions/admin/utxos.go [hunk 8]
```diff
@@ -16,75 +18,44 @@ import (
...
 
-	count, err := reqctx.Engine(c).GetUtxosCount(
-		c.Request.Context(),
-		mappings.MapToMetadata(reqParams.Metadata),
-		conditions,
-	)
-	if err != nil {
-		spverrors.ErrorResponse(c, err, logger)
-		return
+	contracts := make([]*response.Utxo, 0)
```
Comment (chris-4chain) [inline actions/admin/utxos.go:55]: I know that it is moved from the old code.
However, could you reserve the memory to slightly boost it up?
```suggestion
	contracts := make([]*response.Utxo, 0, len(utxos))
```

Could you also, search for similar cases in your changes and do the same?

File: actions/admin/transactions.go [hunk 9]
```diff
@@ -4,81 +4,84 @@ import (
...
-		spverrors.ErrorResponse(c, err, logger)
+		spverrors.ErrorResponse(c, spverrors.ErrCouldNotFindTransaction.WithTrace(err), logger)
 		return
 	}
-
-	contracts := make([]*models.Transaction, 0)
-	for _, transaction := range transactions {
-		contracts = append(contracts, mappings.MapToOldTransactionContractForAdmin(transaction))
+	if transaction == nil {
+		spverrors.ErrorResponse(c, spverrors.ErrCouldNotFindTransaction, logger)
```
Comment (dzolt-4chain) [inline actions/admin/transactions.go:36]: Please also wrap it here

File: actions/admin/transactions.go [hunk 10]
```diff
@@ -4,81 +4,84 @@ import (
...
-func transactionsCount(c *gin.Context, _ *reqctx.AdminContext) {
+func adminSearchTxs(c *gin.Context, _ *reqctx.AdminContext) {
 	logger := reqctx.Logger(c)
-	var reqParams filter.CountTransactions
-	if err := c.Bind(&reqParams); err != nil {
-		spverrors.ErrorResponse(c, spverrors.ErrCannotBindRequest, logger)
+
+	searchParams, err := query.ParseSearchParams[filter.TransactionFilter](c)
+	if err != nil {
+		spverrors.ErrorResponse(c, spverrors.ErrCannotParseQueryParams, logger)
```
Comment (dzolt-4chain) [inline actions/admin/transactions.go:63]: 👆

File: actions/admin/utxos.go [hunk 11]
```diff
@@ -16,75 +18,44 @@ import (
...
 // @Security	x-auth-xpub
 func utxosSearch(c *gin.Context, _ *reqctx.AdminContext) {
 	logger := reqctx.Logger(c)
-	var reqParams filter.AdminSearchUtxos
-	if err := c.Bind(&reqParams); err != nil {
-		spverrors.ErrorResponse(c, spverrors.ErrCannotBindRequest, logger)
+
+	searchParams, err := query.ParseSearchParams[filter.AdminUtxoFilter](c)
+	if err != nil {
+		spverrors.ErrorResponse(c, spverrors.ErrCannotParseQueryParams, logger)
```
Comment (dzolt-4chain) [inline actions/admin/utxos.go:32]: 👆

File: actions/admin/webhooks.go [hunk 12]
```diff
@@ -18,18 +17,19 @@ import (
...
 	if err := c.Bind(&requestBody); err != nil {
-		c.JSON(http.StatusBadRequest, err.Error())
+		spverrors.ErrorResponse(c, spverrors.ErrCannotBindRequest.WithTrace(err), logger)
 		return
 	}
 
 	err := reqctx.Engine(c).SubscribeWebhook(c.Request.Context(), requestBody.URL, requestBody.TokenHeader, requestBody.TokenValue)
 	if err != nil {
-		spverrors.ErrorResponse(c, err, reqctx.Logger(c))
+		spverrors.ErrorResponse(c, spverrors.ErrWebhookSubscriptionFailed, logger)
```
Comment (dzolt-4chain) [inline actions/admin/webhooks.go:32]: 👆

File: actions/admin/xpubs.go [hunk 13]
```diff
@@ -51,68 +52,36 @@ func xpubsCreate(c *gin.Context, _ *reqctx.AdminContext) {
...
 // @Security	x-auth-xpub
 func xpubsSearch(c *gin.Context, _ *reqctx.AdminContext) {
 	logger := reqctx.Logger(c)
-	var reqParams filter.SearchXpubs
-	if err := c.Bind(&reqParams); err != nil {
-		spverrors.ErrorResponse(c, spverrors.ErrCannotBindRequest, logger)
+
+	searchParams, err := query.ParseSearchParams[filter.XpubFilter](c)
+	if err != nil {
+		spverrors.ErrorResponse(c, spverrors.ErrCannotParseQueryParams, logger)
```
Comment (dzolt-4chain) [inline actions/admin/xpubs.go:66]: 👆

File: actions/admin/transactions.go (hunk 9 above)
Comment (chris-4chain) [reply actions/admin/transactions.go:36]: What to Wrap? In this case there is no `error` - just `transaction` is nil which indicates that "Could not find trasnaction" - thats - all.
Looks Good To Me 🙈

File: actions/admin/utxos.go (hunk 8 above)
Comment (ac4ch) [reply actions/admin/utxos.go]: good you pointing this as i was wondering about, but i left this way as most ppl was following this pattern

File: engine/spverrors/definitions.go [hunk 14]
```diff
@@ -95,6 +95,12 @@ var ErrUnknownLockingScript = models.SPVError{Message: "could not recognize lock
 // ErrContactNotFound is when contact cannot be found
 var ErrContactNotFound = models.SPVError{Message: "contact not found", StatusCode: 404, Code: "error-contact-not-found"}
 
+// ErrContactsNotFound is when contacts cannot be found
+var ErrContactsNotFound = models.SPVError{Message: "contacts not found", StatusCode: 404, Code: "error-contacts-not-found"}
+
+// ErrFailedToCountContacts is when contacts cannot be counted
+var ErrFailedToCountContacts = models.SPVError{Message: "failed to count contacts", StatusCode: 400, Code: "error-contacts-count-failed"}
```
Comment (chris-4chain) [inline engine/spverrors/definitions.go:102]:
```suggestion
var ErrFailedToCountContacts = models.SPVError{Message: "failed to count contacts", StatusCode: 500, Code: "error-contacts-count-failed"}
```

Isn't it internal?

File: engine/spverrors/definitions.go [hunk 15]
```diff
@@ -134,6 +140,18 @@ var ErrGettingPKIFailed = models.SPVError{Message: "getting PKI for contact fail
...
+
+// ErrDeleteContact is when deleting contact failed
+var ErrDeleteContact = models.SPVError{Message: "deleting contact failed", StatusCode: 400, Code: "error-contact-deleting-contact-failed"}
+
+// ErrRejectContact is when rejecting contact failed
+var ErrRejectContact = models.SPVError{Message: "rejecting contact failed", StatusCode: 400, Code: "error-contact-rejecting-contact-failed"}
+
+// ErrAcceptContact is when accepting contact failed
+var ErrAcceptContact = models.SPVError{Message: "accepting contact failed", StatusCode: 400, Code: "error-contact-accepting-contact-failed"}
+
```
Comment (chris-4chain) [inline engine/spverrors/definitions.go:154]: I'm not sure if those should be with `StatusCode` 500. 🤔

File: engine/spverrors/definitions.go [hunk 16]
```diff
@@ -142,6 +160,12 @@ var ErrCouldNotFindPaymail = models.SPVError{Message: "paymail not found", Statu
 // ErrPaymailAddressIsInvalid is when the paymail address is NOT alias@domain.com
 var ErrPaymailAddressIsInvalid = models.SPVError{Message: "paymail address is invalid", StatusCode: 400, Code: "error-paymail-address-invalid"}
 
+// ErrDeletePaymailAddress is when the paymail address could not be deleted
+var ErrDeletePaymailAddress = models.SPVError{Message: "deleting paymail address failed", StatusCode: 400, Code: "error-deleting-paymail-address-failed"}
+
+// ErrCreatePaymailAddress is when the paymail address could not be created
+var ErrCreatePaymailAddress = models.SPVError{Message: "creating paymail address failed", StatusCode: 400, Code: "error-creating-paymail-address-failed"}
+
```
Comment (chris-4chain) [inline engine/spverrors/definitions.go:168]: `500` - maybe also here

File: engine/spverrors/definitions.go (hunk 15 above)
Comment (nonus25) [reply engine/spverrors/definitions.go:154]: ok, let keeps 500 for failed actions

File: actions/admin/transactions.go (hunk 9 above)
Comment (dzolt-4chain) [reply actions/admin/transactions.go:36]: Right, misread the line above 🙈

**Review by dzolt-4chain — APPROVED** [review]

**Review by chris-4chain — APPROVED** [review]

**Pull request merged**